web/src/beta/lib/core/mantle/data/gpx.ts

```diff
@@ -14,111 +13,547 @@ export async function fetchGPXfile(
-  const lineStringFeatures = trackPoints.reduce(makeLineStringFeature, []);
-  return lineStringFeatures;
-}
+  return parseGPXWithCustomParser(gpxSource, parseMethod);
+};
 
-type TrackPoint = {
-  "@_lat": string;
-  "@_lon": string;
-  ele: number;
-  time: string;
-  // course: 23.26,
-  // speed: 0.63,
-  // fix: '3d',
-  // sat: 5,
-  // hdop: 1.4
+const parseGPXWithCustomParser = (
+  gpxSource: string,
+  parseGPXToXML: (gpxSource: string) => Document | null,
+) => {
+  // Parse the GPX string using the given parse method
+  const parsedSource = parseGPXToXML(gpxSource);
+
+  // Verify that the parsed data is present
+  if (parsedSource === null) return [null, new Error("Provided parsing method failed.")];
+
+  const output: ParsedGPXInputs = {
+    xml: parsedSource,
+    metadata: {
+      name: "",
+      description: "",
+      time: "",
+      author: null,
+      link: null,
+    },
+    waypoints: [],
+    tracks: [],
+    routes: [],
+  };
+
+  const metadata = output.xml.querySelector("metadata");
+  if (metadata !== null) {
+    // Store the top level elements of the metadata
+    output.metadata.name = getElementValue(metadata, "name");
+    output.metadata.description = getElementValue(metadata, "desc");
+    output.metadata.time = getElementValue(metadata, "time");
+
+    // Parse and store the tree of data associated with the author
+    const authorElement = metadata.querySelector("author");
+    if (authorElement !== null) {
+      const emailElement = authorElement.querySelector("email");
+      const linkElement = authorElement.querySelector("link");
+
+      output.metadata.author = {
+        name: getElementValue(authorElement, "name"),
+        email:
+          emailElement !== null
+            ? {
+                id: emailElement.getAttribute("id") ?? "",
+                domain: emailElement.getAttribute("domain") ?? "",
+              }
+            : null,
+        link:
+          linkElement !== null
+            ? {
+                href: linkElement.getAttribute("href") ?? "",
+                text: getElementValue(linkElement, "text"),
+                type: getElementValue(linkElement, "type"),
+              }
+            : null,
+      };
+    }
+
+    // Parse and store the link element and its associated data
+    const linkElement = queryDirectSelector(metadata, "link");
+    if (linkElement !== null) {
+      output.metadata.link = {
+        href: linkElement.getAttribute("href") ?? "",
+        text: getElementValue(linkElement, "text"),
+        type: getElementValue(linkElement, "type"),
+      };
+    }
+  }
+
+  // Parse and store all waypoints
+  const waypoints = Array.from(output.xml.querySelectorAll("wpt"));
+  for (const waypoint of waypoints) {
+    const point: Waypoint = {
+      name: getElementValue(waypoint, "name"),
+      symbol: getElementValue(waypoint, "sym"),
+      latitude: parseFloat(waypoint.getAttribute("lat") ?? "") || 0,
```

Comment:
```suggestion
      latitude: parseFloat(waypoint.getAttribute("lat") ?? 0),
```
I think we should do like this to avoid confusing.

Reply:
Oh Yeah, Apologies for my ignorance.